Device Password fallback when Touch-ID devices are unavailable

findus opened the pull request:
## Device Password fallback when Touch-ID devices are unavailable

I often use my macbook docked with the lid closed, so I cannot use the fingerprint sensor, and I got rather tired reentering my whole database password every time after unlocking my session.

This change adds the "kSecAccessControlDevicePasscode" flag to the AccessControlFlags, so it is possible to use the device password to unlock the database again.

~This feature is disabled by default and can be enabled in the settings menu.~

* Fixes #8105

## Screenshots
https://github.com/user-attachments/assets/529425a1-18a9-4c84-83ad-9ba958eddac7

![CleanShot 2024-10-25 at 11 52 24@2x](https://github.com/user-attachments/assets/ed8f2310-83bc-415c-b005-0a6619a15c33)


- Touch-ID Hardware present, database locked
- Unlock DB with Touch-ID sensor
- Lid gets closed(no hardware available anymore)
- Lock DB again
- Password fallback kicks in after pressing unlock again

## Type of change
- ✅ New feature (change that adds functionality)

findus committed:
Added kSecAccessControlDevicePasscode to accessControlflags when feature is enabled in settings and no biometrics are available

findus committed:
Included check in isAvailable method

findus committed:
Added toggle button in settings to disable fallback (default: disabled)

findus committed:
Added config check to appropriate method

findus committed:
Use OR to be able to use either biometry or password, if touchid is unavailable

findus committed:
Renamed dialog setting text
translation

findus committed:
i8n update

findus committed:
Clang format

findus committed:
Formatting fixes

droidmonkey commented:
Oh neat, didn't even know that was an option. I don't think we need a special setting to toggle it though. Recommend removing that part and just making this part of the default security control options.

findus committed:
Removed TouchID / WatchID presence check on key write.

Whit this TouchID/WatchID is still possible when DB gets unlocked while no TouchID hardware is present.

findus commented:
@droidmonkey ok then I'll remove the settings menu entry again.

Having this default fallback, that is available all the time, we could also get rid of the isWatchIDAvailable/isTouchIdAvailable checks altogether.

Doing this TouchID/WatchID unlock would always be possbile, even when you open the database for the first time with no biometric/apple watch hardware present and open the lid later.

findus committed:
Removed setting entry again to toggle password fallback

findus committed:
Removed watchID and touchID availability checks

findus committed:
Clang Format

findus committed:
Merge branch 'develop' into feature/touchid-password-fallback

findus committed:
Rephrased comments

findus committed:
Make device passcode unlock always present

findus commented:
Oh, its probably saver to keep all checks in place and add another one with https://developer.apple.com/documentation/localauthentication/klapolicydeviceownerauthentication, I'll do that later this day

findus committed:
Revert "Removed watchID and touchID availability checks"

This reverts commit 0e80d007

findus committed:
Additional check if TouchID is enrolled.

With that we can add the "kSecAccessControlBiometryCurrentSet" Flag even though Biometry is unavailable due to closed lid or unpaired keyboard.

Adding this flag when TouchID is not enrolled results in an error when trying to save the secret.

The kSecAccessControlWatch Flag for apple watch compatibility does not have this limitation.

With that we can also offer quick unlock with only apple watch or password

findus commented:
It was not as easy as I thought, the kSecAccessControlBiometryCurrentSet flag is only allowed when TouchID is enrolled, adding the flag while its not enrolled results in an error when trying to save the secret to the keychain.

But it is possible to catch the specific "is not enrolled" error, and only skip adding the flag when this one occurs.

The kSecAccessControlWatch seems to not have this constraint, but I need to test it a little more, same as my current implementation for TouchID.

findus commented:
I tested i a little more and it seems to be pretty reliable, the only dangerous path is trying to save the key to the enclave with kSecAccessControlBiometryCurrentSet or kSecAccessControlTouchIDCurrentSet as the saving logic is throwing an error when having no fingers scanned, but that path is covered with the new touchidEnrolled check.

With this new logic it is now possible to:

- Use TouchID in this constellations:
  - Docked Laptop, Lid Closed, no Magic Keyboard with TouchID paired -> Open Database -> Lock DB -> Open Lid -> TouchID available
  - Use WatchID Unlock even if TouchID is disabled
  - Use Password unlock  exclusively when no other biometric or companion hardware is present
  - Use WatchID all the time even when pairing/unlocking the watch after unlocking Keepass for the first time

But I dont have an iMac or Mac Mini to test what happens on devices with no integrated TouchID, I tested this constellation in a qemu vm and this case also got handled by the kLAErrorBiometryNotEnrolled error, but I would feel saver if it could get tested with actual imac/mac mini hardware once.

I also could get my hands on a magic keyboard on monday to test that aswell at least once.

findus committed:
Removed config import

droidmonkey commented:
Awesome thank you @findus, this will greatly improve usability of quick unlock. Also helps when I finish implementing "remembered" quick unlock.

findus committed:
fallback to quick unlock without touchid if saving key fails with selected flags, might fix quick unlock on a hackintosh

findus committed:
more documentation to why this logic is necessary

findus committed:
fixed compile error

findus commented:
I've set up a Hackintosh yesterday to check what happens on devices with no TouchID present at all, there the "isEnrolled" check does not work because the API returns LAErrorBiometryNotAvailable, which kinda sucks because now this error is used for two states:

- No Hardware present at all (Setkey fails all the time when the TouchID Flag is used)
- Hardware not present because lid closed/magickeyboard unpaired (SetKey only fails when no fingers enrolled)

To "fix" this I just call setKey a second time without the TouchID/Biometrics flag when the first call fails, that makes it a little more robust.

findus committed:
Botan: always scrub memory, also after failed save attempt

findus commented:
I got my hand on an iMac and I tested the behavior if there is no TouchID hardware present, and what happens if you connect a magic keyboard with sensor to it.

It kinda behaves like on a Hackintosh, only the error message is different, here "kLAErrorBiometryNotPaired" is returned the first time setKey is called, the 2nd fallback call also enables us to use WatchId/DeviceCode.

After you pair the keyboard MacOS guides you through the fingerprint setup process, after that its the same behavior as with a macbook.

After you Unpair the keyboard again and delete your Fingerprints, the API returns "kLAErrorBiometryNotEnrolled" instead of "kLAErrorBiometryNotPaired" 🤔.

But in general its pretty stable now, I used this build for about a week now and haven't encountered anything weird.

droidmonkey commented:
Awesome and thank you for the comprehensive testing! Will merge this shortly.

droidmonkey commented on src/quickunlock/TouchID.mm:
```diff
@@ -332,13 +363,70 @@ inline CFMutableDictionaryRef makeDictionary() {
+
+      bool canAuthenticate = [context canEvaluatePolicy:policyCode error:&error];
+      [context release];
+       //TODO: check if this is the correct error message
```
Did we satisfy this TODO?

findus replied:
We could also get rid of this check entirely now as we call setKey twice in case of an error

droidmonkey replied:
Sounds good

findus committed:
remove enrolled check again, not needed anymore, we just call seKey a second time when first attempt fails with TouchID flag

droidmonkey commented:
Thank you for the contribution!

droidmonkey commented:
This is going to require backporting all the other quick unlock changes that were made if we want to include this in the 2.7.10 release. @phoerious what are you thoughts?
